firstvoices/backend/views/immersion_label_views.py

```diff
@@ -0,0 +1,155 @@
+            404: OpenApiResponse(description=doc_strings.error_404),
+        },
+        parameters=[
+            site_slug_parameter,
+            key_parameter,
+        ],
+    ),
+    partial_update=extend_schema(
+        description="Edit an immersion label in the specified site. Any omitted fields will be unchanged.",
+        responses={
+            200: OpenApiResponse(
+                description=doc_strings.success_200_edit,
+                response=ImmersionLabelDetailSerializer,
+            ),
+            400: OpenApiResponse(description=doc_strings.error_400_validation),
+            403: OpenApiResponse(description=doc_strings.error_403),
+            404: OpenApiResponse(description=doc_strings.error_404),
+        },
+        parameters=[
+            site_slug_parameter,
+            key_parameter,
+        ],
+    ),
+    destroy=extend_schema(
+        description="Delete an immersion label from the specified site.",
+        responses={
+            204: OpenApiResponse(description=doc_strings.success_204_deleted),
+            403: OpenApiResponse(description=doc_strings.error_403),
+            404: OpenApiResponse(description=doc_strings.error_404),
+        },
+        parameters=[site_slug_parameter, key_parameter],
+    ),
+    all=extend_schema(
+        description="Returns a mapping of immersion label keys to their corresponding dictionary entry titles.",
+        responses={
+            200: inline_serializer(
+                name="ImmersionLabelMap",
+                fields={
+                    "key": serializers.CharField(),
+                    "dictionary_entry_title": serializers.CharField(),
+                },
+            ),
+            403: OpenApiResponse(description=doc_strings.error_403),
+            404: OpenApiResponse(description=doc_strings.error_404_missing_site),
+        },
+        parameters=[site_slug_parameter],
+    ),
+)
+class ImmersionLabelViewSet(
+    SiteContentViewSetMixin, FVPermissionViewSetMixin, ModelViewSet
+):
+    """
+    API endpoint that allows immersion labels to be viewed and edited.
+    """
+
+    lookup_field = "key"
+    serializer_class = ImmersionLabelDetailSerializer
+    permission_type_map = {
+        **FVPermissionViewSetMixin.permission_type_map,
+        "all": None,
+    }
+
+    def get_queryset(self):
+        site = self.get_validated_site()
+        return ImmersionLabel.objects.filter(site__slug=site[0].slug).select_related(
+            "site",
+            "site__language",
+            "created_by",
+            "last_modified_by",
+            "dictionary_entry",
+        )
+
+    @action(detail=False, methods=["get"])
+    def all(self, request, *args, **kwargs):
+        """
+        Returns a mapping of immersion label keys to their corresponding dictionary entry titles.
+        """
+        site = self.get_validated_site()
+        immersion_labels = ImmersionLabel.objects.filter(site__slug=site[0].slug)
```

Comment:
Since this is a custom method, you need to apply permissions yourself. Easiest way is to add `.visible(request.user)` to this ImmersionLabel query